File: frontend/nutritheous_app/lib/models/user_profile_request.dart
import 'package:json_annotation/json_annotation.dart';
import 'activity_level.dart';
import 'sex.dart';

part 'user_profile_request.g.dart';

/// Request for updating user profile (matches UserProfileRequest from API)
@JsonSerializable()
class UserProfileRequest {
  final int? age;

  @JsonKey(name: 'heightCm')
  final double? heightCm;

  @JsonKey(name: 'weightKg')
  final double? weightKg;

  final Sex? sex;

  @JsonKey(name: 'activityLevel')
  final ActivityLevel? activityLevel;

  const UserProfileRequest({
    this.age,
    this.heightCm,
    this.weightKg,
    this.sex,
    this.activityLevel,
  });

  factory UserProfileRequest.fromJson(Map<String, dynamic> json) =>
      _$UserProfileRequestFromJson(json);

  Map<String, dynamic> toJson() => _$UserProfileRequestToJson(this);
}

File: backend/src/main/java/com/nutritheous/dto/UserProfileRequest.java
package com.nutritheous.dto;

import com.nutritheous.auth.ActivityLevel;
import com.nutritheous.auth.Sex;
import jakarta.validation.constraints.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * Request DTO for updating user profile information.
 * All fields are optional - only provided fields will be updated.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class UserProfileRequest {

    @Min(value = 10, message = "Age must be at least 10")
    @Max(value = 150, message = "Age must be at most 150")
    private Integer age;

    @DecimalMin(value = "30.0", message = "Height must be at least 30 cm")
    @DecimalMax(value = "300.0", message = "Height must be at most 300 cm")
    private Double heightCm;

    @DecimalMin(value = "1.0", message = "Weight must be at least 1 kg")
    @DecimalMax(value = "500.0", message = "Weight must be at most 500 kg")
    private Double weightKg;

    private Sex sex;

    private ActivityLevel activityLevel;
}
